### PR TITLE
Snapshotting AggregateRoot

### PR DESCRIPTION
**Pull request by riasvdv**
This implements snapshotting of AggregateRoots
- Adds a `snapshot()` method to the AggregateRoot which creates a new Snapshot
- AggregateRoots track a `version` number, which increments each time an event is applied
- Persisting the events persists them with the version number
- The AggregateRoot has a `getState` and `useState` function to determine the state that is saved and retrieved from the snapshot, by default all properties on the class are stored in the snapshot

- When initializing an AggregateRoot, we check if there is a snapshot, if there is we load the state first, then retrieve and reconstitute from all events that happened after the snapshot.
- All public properties on an aggregate are stored as the state of the snapshot, in the tests this is the `$balance` property of the `AccountAggregate`, when restoring a snapshot, we set them back.

- We should also credit https://github.com/EventSaucePHP/EventSauce as the implementation is based on how EventSauce does it.

This will require a new major version, as some interfaces and a migration have changed, steps to upgrade:
- Add an `aggregate_version` property to the `stored_events` table `$table->unsignedInteger('aggregate_version')->nullable();`
- Republish the migrations or copy the `create_snapshots_table` migration
- The `StoredEventRepository` interface has a new method `retrieveAllAfterVersion` that you must implement if you have a custom repository
- The `StoredEventRepository` now accepts an `aggregateVersion` parameter in the `persist` and `persistMany` methods

### CODE REVIEW

File: src/AggregateRoot.php [hunk 1]
```diff
@@ -67,6 +84,16 @@ public function getRecordedEvents(): array
         return $this->recordedEvents;
     }
 
+    private function state(): array
+    {
+        $class = new ReflectionClass($this);
+
+        return collect($class->getProperties(ReflectionProperty::IS_PUBLIC))
```
Comment (freekmurze) [inline src/AggregateRoot.php]: Does this mean that if you want to use snapshots, all properties you want to use should be public?

File: src/AggregateRoot.php [hunk 2]
```diff
@@ -78,7 +105,23 @@ private function getAndClearRecordedEvents(): array
 
     private function reconstituteFromEvents(): AggregateRoot
     {
-        $this->getStoredEventRepository()->retrieveAll($this->aggregateUuid)
+        $storedEventRepository = $this->getStoredEventRepository();
```
Comment (freekmurze) [inline src/AggregateRoot.php]: Maybe we should try to find and use a snapshot first. 

In either case we could call `retrieveAllAfterVersion`. If there was no snapshot that function will (or should) start getting all events, if there was a snapshot it should only get events after the snapshot (I guess that already works now)

File: stubs/create_snapshots_table.php.stub [hunk 3]
```diff
@@ -0,0 +1,20 @@
+<?php
+
+use Illuminate\Support\Facades\Schema;
+use Illuminate\Database\Schema\Blueprint;
+use Illuminate\Database\Migrations\Migration;
+
+final class CreateSnapshotsTable extends Migration
```
Comment (freekmurze) [inline stubs/create_snapshots_table.php.stub]: Let's just drop all usage of `final`.

Comment (riasvdv) [reply stubs/create_snapshots_table.php.stub]: I'll do this in a separate PR

File: src/AggregateRoot.php [hunk 4]
```diff
@@ -67,6 +84,23 @@ public function getRecordedEvents(): array
         return $this->recordedEvents;
     }
 
+    protected function getState(): array
```
Comment (freekmurze) [inline src/AggregateRoot.php]: Nice!

**Comment by freekmurze** [comment]
Very nice work @riasvdv! I'll already merge this to v3. Could you also add docs for this on that branch?

**Comment by warren32** [comment]
This is great thank you @riasvdv

**Pull request merged**